### ISSUE TITLE
cabal new-build does not track files added via TH’s addDependentFile

### ISSUE THREAD

**Issue by cocreature** [haskell/cabal#4746]
### Steps to reproduce
* Download https://gist.github.com/cocreature/d07ef6f35f7bd9043db50794fe31e977. 
* Build using `cabal new-build`
* Edit `fileinput`
* Try rebuilding using `cabal new-build`.

### Expected behavior
`cabal new-build` rebuilds because `fileinput` has changed.

### Actual behavior
`cabal new-build` exits without rebuilding after saying `Up to date`.

This issue affects both `cabal-2.0` as well as the master branch.

**Comment by phadej** [comment]
One hack solution is to watch `extra-source-files` too. But that won't solve more dynamic (e.g. what `inline-c` uses?) stuff.

**Comment by 23Skidoo** [comment]
BTW, is there a way to make `ghc --make` always rerun a TH function? E.g. if I'm doing some kind of I/O at compile-time, like reading from an environment variable.

**Comment by phadej** [comment]
@23Skidoo that would force GHC to always build that module, that doesn't sound like a very good idea.

**Comment by 23Skidoo** [comment]
Yep, but sometimes you want that in a build system. For example, Shake rules can be marked `alwaysRerun`.

Re: `extra-source-files`, that field is package-global and commonly used for things like `README` and `changelog`. So we'd need to allow per-component `extra-source-files` to avoid unnecessary recompilation

**Comment by 23Skidoo** [comment]
An alternative way to fix this is to get the list of files to be monitored from `ghc --make` via `-ddump-module-graph`.

**Comment by phadej** [comment]
Does `-ddump-module-graph` dump `addDependentFile` & co stuff?

**Comment by 23Skidoo** [comment]
@phadej If you click that link, you'll see that I asked that exact question there.

**Comment by cocreature** [comment]
Given that this feature will only be in `GHC 8.4` I don’t think that’s a reasonable solution at this point.

**Comment by 23Skidoo** [comment]
@cocreature With old versions we can simulate that feature with `ghc -M` (which will be slower because we'll have to do dependency resolution twice - once for `ghc -M` and once for `ghc --make`, but for legacy GHC versions we can afford to not care). BTW, `ghc -M` does support `addDependentFile` dependencies:

```
# DO NOT DELETE: Beginning of Haskell dependencies
Data/FileEmbed.so : Data/FileEmbed.hs
Main.so : Main.hs
Main.so : Data/FileEmbed.shi
# DO NOT DELETE: End of Haskell dependencies
```

**Comment by hvr** [comment]
@23Skidoo btw, we can avoid the slow-legacy fallback because we know when TH is needed (we already require `TemplateHaskell` to be declared in `*-extensions` in order to enable other logic specific to TH). So I think that's a workable fallback.

**Comment by hvr** [comment]
Unfortunately we not only missed the boat for GHC 8.4, but most likely also for GHC 8.6 to get supporting features into GHC... as https://phabricator.haskell.org/D3898 got stuck on a decision which design path to go down... :-/

**Comment by kaldonir** [comment]
We ran into the same issue today. Is there any progress here?

**Comment by phadej** [comment]
@kaldonir, no. As far as I know no-one is actively working on adding necessary features to GHC (e.g. https://github.com/ghc-proposals/ghc-proposals/pull/245 or something similar so the compiler could tell Cabal what extra dependencies exist).

Alternatively or additionally `cabal` could track `extra-source-files`, which is imprecise yet better than nothing. Again, I'm not aware anybody is working on that either.

**Comment by parsonsmatt** [comment]
So [parsing the `.dump-hi` files was discouraged a while ago](https://github.com/haskell/cabal/issues/4746#issuecomment-411554751). I'm curious if there's any objection to just doing that, since [it's what `stack` has been doing since 2015 with one bug report in that time](https://github.com/commercialhaskell/stack/commit/03dfaa88d461647cc084efa656b18abb73886bab#diff-f877da3bc1f76cae2cd5094f15dd2d03e2af5be4a044da9af28bb1bf27cb33a8R1289-R1293). 

If this is an acceptable solution then I can allocate some time and resources to implementing it.

Thanks!

**Comment by Ericson2314** [comment]
GHC should be able to give cabal this info in a machine consumable cormat.

But frankly I don't really like the design of `addDependentFile` because it precludes proper sandboxing — it's very dynamic sort of dependency.

**Comment by parsonsmatt** [comment]
> GHC should be able to give cabal this info in a machine consumable cormat.

Agreed, but it hasn't started doing that in the 4 years that this ticket has been open. I'm interested in solving it now.

**Comment by gbaz** [comment]
@parsonsmatt for solving now, would simply tracking `extra-source-files` as suggested above suffice? That should be straightforward to implement.

**Comment by parsonsmatt** [comment]
We've got ~176 files to track (and counting), and manually tracking them in the cabal file would be a major nuisance. But patching the wildcard behavior would suffice, though I do still think this should be a seamless process.

**Comment by Ericson2314** [comment]
GHC's driver is being overhauled at the moment, which will help with this and other dependency issues.

**Comment by Mikolaj** [comment]
Yes, I remember tracking extra source files started working again, recently. Is there a ticket about adding wildcards to `extra-source-files`? The closest I could find was https://github.com/haskell/cabal/issues/7290.

**Comment by Ericson2314** [comment]
Yeah tracking `extra-source-files` is definitely the right short-term solution.

**Comment by gbaz** [comment]
I have a pr now for glob expanding extra-src-files and am moving the remainder of the ticket to blocked on ghc, because I think we _really_ don't want to hack around ghc proprietary formats to try to extract this information in any more sophisticated way.

**Comment by liftM** [comment]
Is there an easy way to instrument `addDependentFile` calls so we can figure out what files to add to `extra-source-files` as a workaround?

**Comment by gbaz** [comment]
The type is `FilePath -> Q ()` -- you can wrap it in any other function you want including IO actions lifted to `Q` to output the data however you desire.

**Comment by parsonsmatt** [comment]
If you're able to operate in an abstract monad, like `Quasi m => m _`, then you can provide a `newtype` around `Q` and define an instrumented [`qAddDependentFile`](https://www.stackage.org/haddock/lts-18.16/template-haskell-2.16.0.0/Language-Haskell-TH-Syntax.html#v:qAddDependentFile). But this won't work for anything hardocded to `Q`, which is quite a lot of the API.

**Comment by phadej** [comment]
@parsonsmatt it will. [`runQ :: Quasi m => Q a -> m a`](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/Language-Haskell-TH.html#v:runQ) is your friend.

**Comment by liftM** [comment]
Yeah, this works for my own code calling `addDependentFile`. What I'm really trying to do is figure out how to intercept and instrument `addDependentFile` calls from external TH libraries that I depend on (e.g. `file-embed`). It seems that this is not possible.

**Comment by jberryman** [comment]
Can confirm in cabal `3.6.2.0` when I add the argument to addDependentFile to extra-source-files, addDependentFile seems to work properly (and an unchanged module isn't recompiled)

**Comment by michaelpj** [comment]
This seems to have broken again in 3.8.1, are we sure that https://github.com/haskell/cabal/pull/7608 is correct?

**Comment by SamirTalwar** [comment]
We saw the same problem in https://github.com/hasura/graphql-engine, but unfortunately I couldn't reproduce it with a little project, so I figured we were doing something wrong. Interesting to know we're not the only ones.

For now, we have downgraded to Cabal 3.6 in CI.

**Comment by gbaz** [comment]
Well 3.6.2.0 was just a backport of 7608, so if that was correct, I don't see how it would have broken in main unless there was an unrelated regression? Can you expand with a reproducer, or at least a more cogent statement of what the expected vs actual behavior is?

**Comment by michaelpj** [comment]
I haven't tried to minimize yet. It's large, but here's our actual reproducer: https://github.com/input-output-hk/plutus

```
cabal build plutus-core:plutus-core
# edit plutus-core/cost-model/data/builtinCostModel.json>
cabal build plutus-core:plutus-core
# observe no recompliation
```

We've confirmed that this works on cabal 3.6.2, and not on 3.8.1. I've tried a few semi-random things to try and get cabal to notice the change:
1. Move the file to the package root to avoid directory separators in the pattern.
2. Remove all the other `extra-source-files`
3. Explicitly add a call to `addDependentFile <path>` to the relevant module.
4. Changing to use a wildcard match on the file extension
5. Bumping the `cabal-version` of the cabal file

> Well 3.6.2.0 was just a backport of 7608, so if that was correct, I don't see how it would have broken in main unless there was an unrelated regression? 

Okay, well we have it working on 3.6.2 so my finger-pointing at 7608 must be wrong :) Then I'm quite baffled!!

**Comment by michaelpj** [comment]
One more failed experiment:
6. Add the package directory to the `extra-source-file` path in case there was a CWD mixup and it was looking from the project root somehow.

**Comment by michaelpj** [comment]
(Perhaps as a tangent, but seeing the discussion [here](https://github.com/haskell/cabal/pull/7608#discussion_r699616666): maybe I don't want cabal to _die_ if an `extra-source-file` line matches nothing, but I sure would like a _warning_. It almost certainly means I've made a mistake as the user!)

**Comment by Mikolaj** [comment]
@michaelpj: may I ask you to open a new ticket, copy the reproduction info and mark it a regression in 3.8? That would be high priority to investigate before 3.10. I'd at least make sure the https://github.com/haskell/cabal/pull/7608 commit has not been lost on 3.8 branch and on current master.

Re warning when glob is empty, that also merits a new ticket, perhaps even a newcomer one.

**Comment by michaelpj** [comment]
Will do.

**Comment by michaelpj** [comment]
https://github.com/haskell/cabal/issues/8632
https://github.com/haskell/cabal/issues/8634

I don't have labeling permissions, so I can't label them, sorry.

**Comment by ulysses4ever** [comment]
@Mikolaj we should give the permissions to Michael, I think

**Comment by Mikolaj** [comment]
No way, @michaelpj, I've just checked and you weren't even a lowly Read-level member of the project according to the Settings page. I've invited you to the respectable Triage level. Could you try?

**Comment by michaelpj** [comment]
done

**Comment by Mikolaj** [comment]
Perfect. Thank you.

**Comment by hsyl20** [comment]
Stack's code for parsing `.hi` files is in a separate library: https://github.com/commercialhaskell/hi-file-parser. I've recently updated it to support 9.4.5 and 9.6.1. Cabal could use it to detect dynamic dependencies per module. Th dependency on RIO could easily be removed.

**Comment by gbaz** [comment]
Thinking about hi based tracking I think there's a potential issue: if files are not in the .cabal manifest they still won't get packaged up by an sdist. So that means that an `sdist` of a package that works locally with some "dynamic" tracking will fail if packaged up. I think that's not a good situation. As it stands, I'm not sure why tracking extra-src-files for rebuild along with glob-expansion of them doesn't suffice.

**Comment by KristianBalaj** [comment]
I think this is a related issue I've just created - https://github.com/haskell/cabal/issues/9745

**Comment by NorfairKing** [comment]
I'm very surprised to see this being labeled as "priority:low" given that I consider this a dealbreaking "cannot consider cabal usable until fixed"-bug.

**Comment by Mikolaj** [comment]
@NorfairKing: I've bumped priority. PRs welcome. Actually, design welcome first, IIRC? Especially that it's marked as blocked on GHC, so we either need to create a patch for GHC or devise a special workaround?

**Comment by NorfairKing** [comment]
@Mikolaj I don't understand how this is blocked on GHC when stack has not had this bug since 2015.

**Comment by Mikolaj** [comment]
Me neither, but it's a long time since I looked into this long discussion thread (assuming I understood it then). Do you think things may have changed since then? Would you like to re-assess?

**Comment by TeofilC** [comment]
My understanding is that stack gets this information by parsing binary `.hi` files using (See: https://github.com/commercialhaskell/stack/issues/5134).

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11994 is some movement towards getting a proper interface for exposing the dependency graph from the GHC side. Though I'm not sure if that includes TH dependencies

**Comment by geekosaur** [comment]
> @Mikolaj I don't understand how this is blocked on GHC when stack has not had this bug since 2015.

Only GHC knows the result of running TH (do you really want cabal duplicating enough of GHC that it can run TH itself? Remember that you can do conditionals) so it would have been blocked on GHC exporting that information usefully. Later messages in this thread indicate it's now part of the `.hi` file, so presumably this can proceed now.

**Comment by Mikolaj** [comment]
Thank you for your remarks. 

@phadej has also educated me on this issue just now: 

"
The https://github.com/haskell/cabal/issues/4746 is softly fixed in Cabal too. If you add a file to extra-source-files, it will track its dirtyness. Maybe stack just tracks everything (i.e. overestimates even more). To fix that issue exactly, GHC must tell what files were actually used, so they could be recorded as an exact dependency.
like CPP implementation tell which headers were actually included when you preprocess a source file.
(FWIW, these aren't recorded either, it's just not so obvious)
https://gitlab.haskell.org/ghc/ghc/-/issues/25017
addDependentFile dependencies are at least recorded in the interface files, so if one really wants, then by parsing ghc --show-iface output one can figure those out.
"

So maybe we could look into the `.hi` files. Worth investigating. Volunteers welcome.

**Comment by ulysses4ever** [comment]
> So maybe we could look into the .hi files. 

That's what Sylvain said above: Stack parses hi for this reason (not sure alone or not) with https://github.com/commercialhaskell/hi-file-parser, and we should adopt it.

**Comment by phadej** [comment]
@ulysses4ever 

> https://github.com/commercialhaskell/hi-file-parser, and we should adopt it.

No you shouldn't. see e.g. https://github.com/commercialhaskell/stack/issues/5134.

Stack doesn't some not very good design choices, and that's a good example.

But if you want one more thing to slow down making recent ghc compatible releases, go ahead: AFAICT, `hi-file-parser` doesn't support GHC-9.10, at the very least it's not tested: the last commit is from 7 months ago.

**Comment by mpickering** [comment]
I completely agree with Oleg that depending on "hi-file-parser" is a bad idea.

**Comment by ulysses4ever** [comment]
@phadej @mpickering what are our options?

Reading Mikolaj's post above, I thought that Oleg supports the idea of parsing `.hi` files. Is there anything wrong with `hi-file-parser` in particular or is there anything else that is both better and achievable with our limited resources?

This issue looks important for some amount of users. Feature-parity with stack is a desirable goal too (modulo some things like installing GHC). I wish we could come to an agreement on some way forward for this.

**Comment by phadej** [comment]
>  I thought that Oleg supports the idea of parsing .hi files. 

I didn't say that, and I only implied that one could parse `ghc --show-iface` output, but even that is fragile, and thus bad idea.
